src/openassetio-core/src/utils/Regex.hpp

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2023 The Foundry Visionmongers Ltd
+#pragma once
+#include <cstddef>
+#include <memory>
+#include <optional>
+
+#define PCRE2_CODE_UNIT_WIDTH 8
+#include <pcre2.h>
+
+#include <openassetio/export.h>
+#include <openassetio/typedefs.hpp>
+
+namespace openassetio {
+inline namespace OPENASSETIO_CORE_ABI_VERSION {
+namespace utils {
+
+/**
+ * Regular expression compilation, matching and caching.
+ *
+ * Wraps PCRE2, using its JIT compilation and matching functions.
+ *
+ * Instances of this class are _not_ thread-safe. Use a separate
+ * instance per thread.
+ *
+ * As well as the regex object itself, matches are cached for subsequent
+ * querying.
+ */
+class Regex {
+ public:
+  /**
+   * Container for a regex match.
+   */
+  class Match {
+    struct MatchDataDeleter {
+      void operator()(pcre2_match_data* ptr) { pcre2_match_data_free(ptr); }
+    };
+    using Data = std::unique_ptr<pcre2_match_data, MatchDataDeleter>;
+
+   public:
+    explicit Match(const pcre2_code* code);
+
+    /**
+     * Get the string from a group in the match.
+     *
+     * Warning: no validation is performed. It is assumed the given
+     * group number exists in the match data.
+     *
+     * @param subject Subject string to extract a substring view from.
+     * Should be the same as the subject of the original `match`
+     * call, or at least a string of equal length.
+     *
+     * @param groupNum Regex capture group to extract.
+     *
+     * @return Substring view of the match in the subject string.
+     */
+    [[nodiscard]] StrView group(StrView subject, std::size_t groupNum) const;
+
+    /**
+     * Get the internal match data pointer.
+     *
+     * @return Internal PCRE2 match data.
+     */
+    [[nodiscard]] const Data& data() const { return data_; }
+
+   private:
+    Data data_;
+  };
+
+  /**
+   * Constructor.
+   *
+   * Pre-compiles the regular expression pattern.
+   *
+   * Note that:
+   * - Patterns are case-insensitive.
+   * - `$` matches end of string, not newline.
+   * - `.` matches all characters, including newlines.
+   *
+   * @param pattern Regex pattern.
+   */
+  explicit Regex(StrView pattern);
+
+  ~Regex();
+
+  /**
+   * Check if the regex matches a given subject string.
+   *
+   * Caches the match results for subsequent retrieval in other methods.
+   *
+   * @param subject
+   *
+   * @return `true` if there is a match, `false` otherwise.
+   */
+  [[nodiscard]] std::optional<Match> match(StrView subject) const;
```

Comment:
In be8e6168 I changed the `Regex` class to not hold mutable state, i.e. we allocate a new `pcre2_match_data` with every `match`/`substituteToReduceSize` call.

Timings:

Release build, test cases x10 (=13320), hacked pytest to increase resolution of summary timings to 6dp:

| Per-match (new) | Reused (old) |
|-------------|-------------|
| 2.678623 | 2.667705 |
| 2.568904 | 2.499892 |
| 2.591064 | 2.595332 |

It's pretty indistinguishable.  I also ran with the normal number of test cases (1332) and in Debug mode before I got these Release x10 figures, and taken all together the reused match data block does seem to give slightly faster average results. But its so marginal, and the benefit of thread-safety and `const`-ing all the things seems worth the hit.

Reply:
Thanks for running these numbers, sort of what I expected.
Glad to see statelessness here, as you've outlined, the benefits are immense.